### ISSUE TITLE
[Pal/Linux-SGX] Program crashes on exit from multi-threaded execution

### ISSUE THREAD

**Issue by alexandermerritt** [oscarlab/graphene#824]
I hope I am not conflating multiple problems together, but I am able to verify that two behaviors together cause undesired program exiting or crashing: exiting a program while threads created by main continue to execute.

**Symptom:**

In Linux-SGX, ~15% of the time (like 3 every 20 runs), the program experiences an AEX upon exiting the application and hangs.

This problem arose from testing Golang applications. The runtime spawns threads on creation of the process, even if you do not use any in the implementation (e.g., for garbage collection, etc.). When the program exits, these Golang threads are still around and result in multiple (nested) entries into `_DkExceptionHandler` which results in an AEX and program halt. Golang apps were used on an as-yet-unpublished branch of Graphene used internally at Intel. I've tried replicating it with a C program on the mainline master branch, documented below.

**Root cause:**

Have not yet identified. Not sure next steps to investigate.

**Examples**

On the master branch commit 124a418 compiled to use Linux-SGX you can try this C application, placed in `LibOS/shim/test/native`. You'll need to add `-pthread` to `LDFLAGS` or to the linker command line, and add `sgx.thread_num = 16` to `manifest.template` before building.

```C
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <unistd.h>

const int N = 8;   // total threads
_Atomic int S = 0; // threads which have started
_Atomic int E = 8; // threads which have ended

void* thread(void* arg) {
    const int n = 1<<14;
    for (int i = 0; i < n; i++) {
        void *m = mmap(NULL, 4096, PROT_READ, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0);
        if (m != MAP_FAILED) munmap(m,4096);
        if (i == n/8) S++;
    }
    E--;
    puts("thread done");
    return NULL;
}

int main() {
    pthread_t tids[N];
    for (int i = 0; i < N; i++) {
        if (0 != pthread_create(&tids[i], NULL, thread, NULL))
            return 1;
    }
    while (S < N) ; // wait for threads to be in-progress
    // while (E != 0) ; // wait for threads to complete
    puts("--------------- Done");
    return 0;
}
```

Shows the following errors in the output:

```
(initial output omitted)
[     9] bkeep_unmapped: 0x8097cf000-0x8097d0000
[     6] ---- return from shim_mmap(...) = 0x8097d1000
[     1] ---- shim_write(1,0x8097cb000,21) ... write
[     8] ---- shim_mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[     6] ---- shim_munmap(0x8097d1000,4096) ... munmap
[     9] ---- return from shim_mmap(...) = 0x8097cf000
[     8] bkeep_unmapped: 0x8097ce000-0x8097cf000
--------------- Done
[     9] ---- shim_munmap(0x8097cf000,4096) ... munmap
[     1] ---- return from shim_write(...) = 21
[     8] ---- return from shim_mmap(...) = 0x8097ce000
[     4] bkeep_munmap: 0x8097d0000-0x8097d1000
[     1] ---- shim_exit_group (returning 0)
[     8] ---- shim_munmap(0x8097ce000,4096) ... munmap
[     9] bkeep_mprotect: 0x8097cf000-0x8097d0000
[     5] bkeep_unmapped: 0x8097cd000-0x8097ce000
[     1] now kill other threads in the process
[     1] walk_thread_list(callback=0x80b501f94)
[     5] ---- return from shim_mmap(...) = 0x8097cd000
[     1] Thread 2 killed by signal 9
[     6] bkeep_mprotect: 0x8097d1000-0x8097d2000
[     2] bkeep_mprotect: 0x8097ca000-0x8097cb000
[     7] bkeep_unmapped: 0x8097c9000-0x8097ca000
[     5] ---- shim_munmap(0x8097cd000,4096) ... munmap
[     1] signal_logs[8]: head=0, tail=1 (counter = 1)
[     1] resuming thread 2
[     7] ---- return from shim_mmap(...) = 0x8097c9000
[     5] bkeep_mprotect: 0x8097cd000-0x8097ce000
[     9] bkeep_munmap: 0x8097cf000-0x8097d0000
[     7] ---- shim_munmap(0x8097c9000,4096) ... munmap
[     1] Thread 3 killed by signal 9
[     4] ---- return from shim_munmap(...) = 0
[     1] signal_logs[8]: head=0, tail=1 (counter = 1)
[     5] bkeep_munmap: 0x8097cd000-0x8097ce000
[     9] ---- return from shim_munmap(...) = 0
[     1] resuming thread 3
[     4] ---- shim_mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[     7] bkeep_mprotect: 0x8097c9000-0x8097ca000
[     9] ---- shim_mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
[     5] ---- return from shim_munmap(...) = 0
[     4] bkeep_unmapped: 0x8097d0000-0x8097d1000
[     1] Thread 4 killed by signal 9
[     5] ---- shim_mmap(0x0,4096,PROT_READ,MAP_PRIVATE|MAP_ANON,-1,0) ... mmap
*** An unexpected AEX vector occurred inside PAL. Exiting the thread. ***
(vector = 0x2, type = 0x0 valid = 0, RIP = +0x00019ba8)
rax: 0x7fd7fbd1aea0 rcx: 0x80b993ba8 rdx: 0x80b993ba8 rbx: 0x7fd7fbd1aea0
rsp: 0x800610860 rbp: 0x80ff38d88 rsi: 0x00000005 [     1] signal_logs[8]: head=0, tail=1 (counter = 1)
[     3] bkeep_unmapped: 0x8097cf000-0x8097d0000
[     4] ---- return from shim_mmap(...) = 0x8097d0000
[     1] resuming thread 4
(more output omitted)
```

Many of the register values look strange. RIP is `19ba8:       48 8b 0e                mov    rcx,QWORD PTR [rsi]` in function `_wordcopy_fwd_aligned`, accessing something in address 0x5.

Another example error

```
*** An unexpected AEX vector occurred inside PAL. Exiting the thread. ***
(vector = 0x2, type = 0x0 valid = 0, RIP = +0x0000e7e8)
rax: 0x00000000 rcx: 0x80b988790 rdx: 0x00000001 rbx: 0x00000014
rsp: 0x8019fcaf0 rbp: 0x7f866b3f0ec0 rsi: 0x00000001 rdi: 0[     1] signal_logs[8]: head=0, tail=1 (counter = 1)
[     7] bkeep_unmapped: 0x80afd6000-0x80afd7000
[     9] ---- return from shim_munmap(...) = 0
x7f866dc1a1d0
r8 : 0x00000001 r9 : 0x00000000 r10: 0x7f866dc1a1d0 r11: 0x00000001
r12: 0x7f866b3f0ec0 r13: 0x00000014 r14: 0x00000001 r15: 0x00000000
rflags: 0x00010297 rip: 0x80b9887e8
```

RIP is `e7e8: f0 41 ff 4e 08          lock dec DWORD PTR [r14+0x8]` in function `_DkMutexLockTimeout` which seems to be accessing a location in page zero?

**Possible Related Issues**

#517 and #163

**Comment by dimakuv** [comment]
Will this PR help: https://github.com/oscarlab/graphene/pull/641? It will be finished and merged soon.

**Comment by yamahata** [comment]
Nice report. because you have already a test, can you please upload as PR to have test in shim/test/regression? so test can be run easilty.

**Comment by alexandermerritt** [comment]
Submitted PR #846 for testing

**Comment by dimakuv** [comment]
I've closed #846 (see explanation there) and thus closing this issue as well. Feel free to re-open it if it is still useful.
